### PR TITLE
Wallet: Add foreign_outputs metadata to support CoinJoin transactions

### LINKED ISSUE #14136: rpc: Wrong `gettransaction` info for a coinjoin
Take a look at this transaction:

https://testnet.smartbit.com.au/tx/947ddb7d22b0c7e8edd3a56a92b8ec771aa8b18667ef508270f928b14821e924

I've double checked with `getaddressinfo` and only the first input is mine, and the second output is mine. So the transaction looks like

Inputs: [1.02 BTC mine,  1.1 BTC not-mine]
Outputs:  [0.10481803 BTC not-mine, 2.01 BTC mine]

So clearly this transaction makes me  0.99 BTC richer.  (-1.02 + 2.01)

So now let's look at what `gettransaction` returns:

```
{
  "amount": -0.10481803,
  "fee": 1.09481803,
  "confirmations": 0,
  "trusted": false,
  "txid": "947ddb7d22b0c7e8edd3a56a92b8ec771aa8b18667ef508270f928b14821e924",
  "walletconflicts": [
  ],
  "time": 1535929790,
  "timereceived": 1535929790,
  "bip125-replaceable": "no",
  "details": [
    {
      "address": "2N8qCRsKMJ3qNYo8LRb9kLpy3qiigAuedMT",
      "category": "send",
      "amount": -0.10481803,
      "vout": 0,
      "fee": 1.09481803,
      "abandoned": false
    },
    {
      "address": "2MzFcfPF56Jp6BC3FJk6yXwCY6FzhnbGt3o",
      "category": "send",
      "amount": -2.01000000,
      "label": "",
      "vout": 1,
      "fee": 1.09481803,
      "abandoned": false
    },
    {
      "address": "2MzFcfPF56Jp6BC3FJk6yXwCY6FzhnbGt3o",
      "category": "receive",
      "amount": 2.01000000,
      "label": "",
      "vout": 1
    }
  ],
  "hex": "020000000001022eb1b5698216f56eea3770c949a5d9555dcd1062571cc3113cae68132a5ba5660100000017160014f67a19d36a915d7e0360d3a66d1f055c7bd34687feffffffe165f072311e71825b47a4797221d7ae56d4b40b7707c540049aee43302448a401000000171600142fa1c410210f8228467e53f450c9e4e56337895ffeffffff028bf09f000000000017a914aaf6ba68cf6a2369445f119596adb84412d28d3a874004fb0b0000000017a9144cdbced31bab78d8a248204a961a4ad475d6b48a87024730440220724dbbdc278315ed367036b163749702c7810730dc2b17ba3226f72c7d0078ab02207945ffb8ffc78bcefa207ac8e172983b5c3d91f64c3cb869cdcd288e6c1d85610121037426b52590172ce72df80829e4a9f7530224306399c399080fc917d19ae44a72024730440220168c025d48f3ee4d626d37a85c92a75e778649e966d8f008caeed688cb6da73902202e00fb70c80e0ba591a12a8feb067cf3216eb054ac5059d0f38ce805bbdd478701210213be1638b4620ccacf4e1bec9748c3fc27561aa7169dbd1639be00bb20236ab800000000"
}
```

Note the "fee" thing it has is pretty confused. The actual fee is: 0.00518197 BTC

Running on the 0.17   branch if it makes a difference.

### PR DESCRIPTION
**Pull request by luke-jr**
Fixes #14136

Does not fix GUI (can be done in GUI-side PR afterward)

### CODE REVIEW

**Comment by 1440000bytes** [comment]
Concept ACK.

Will test today.

> Edit: It's not entirely clear at this point how to avoid older versions from opening a wallet with this metadata. Advice?

Not sure about this. If there is any error in old wallets that should work. Users with such issues can post on stackexchange, reddit etc. 

Can also add information in release notes.

**Comment by 1440000bytes** [comment]
I don't think this fixes https://github.com/bitcoin/bitcoin/issues/14136

I tried it with my one of my signet wallet that was used for testing of joinstr transaction: `75e490b10b15a6a0422f25ff66ad98ef70390c8fecaac02712705dce8cc3564b`

```
$ bitcoin-cli -rpcwallet="w5" gettransaction 75e490b10b15a6a0422f25ff66ad98ef70390c8fecaac02712705dce8cc3564b
{
  "amount": -0.00583000,
  "fee": 0.00465000,
  "confirmations": 2160,
  "blockhash": "00000007567d87fadb4673d0cb876cb1b7b9dbfedc1d9a0e40af7380d9da8435",
  "blockheight": 104271,
  "blockindex": 1,
  "blocktime": 1660978126,
  "txid": "75e490b10b15a6a0422f25ff66ad98ef70390c8fecaac02712705dce8cc3564b",
  "wtxid": "e4b91ce21961a9e5498e2ff396f5e09fa8c7f3a36ef77abd4207f71855599990",
  "walletconflicts": [
  ],
  "time": 1660976747,
  "timereceived": 1660976747,
  "bip125-replaceable": "no",
  "details": [
    {
      "address": "tb1qhxrp4zl54ul0twtyz0gury5399q7z0kvqqrl6m",
      "category": "send",
      "amount": -0.00116600,
      "vout": 0,
      "fee": 0.00465000,
      "abandoned": false
    },
    {
      "address": "tb1qnnflzn8gwadhwxr46zpxgc8fy429t25w35wah9",
      "category": "send",
      "amount": -0.00116600,
      "vout": 1,
      "fee": 0.00465000,
      "abandoned": false
    },
    {
      "address": "tb1q85ny57lu8g7na23qt7tgf03srksh9x3hrpgkvq",
      "category": "send",
      "amount": -0.00116600,
      "vout": 2,
      "fee": 0.00465000,
      "abandoned": false
    },
    {
      "address": "tb1qe36279395n6crhx3d09cptvnv72chagd336jla",
      "category": "send",
      "amount": -0.00116600,
      "vout": 3,
      "fee": 0.00465000,
      "abandoned": false
    },
    {
      "address": "tb1qkdeccjk32n0muv45xp67kq2pfd3qvfe6wym0ey",
      "category": "send",
      "amount": -0.00116600,
      "vout": 4,
      "fee": 0.00465000,
      "abandoned": false
    }
  ],
  "hex": "02000000000105b4c6a8258717bce1b92f761acf76324bb808b78879ff3ccead1452de1ad5bf0a0000000000ffffffffa8f3781e6929b82fbe8a4e821497932266a2c051b20324baba0d6ae33ffddbf00100000000ffffffff3105dc25c8bd51e921aecda35b35ca0f8ab7239f5d3452c969ce75731fcdc0f00100000000ffffffffc0c6b85c082753edfd21e9fe286f6b62dbfc6cb30d61a92667263f4057f02d170000000000fffffffff9338d880fab8b9904dbdad9c51e737b50ca670c3db9b8df195fb03e18d48cdd0100000000ffffffff0578c7010000000000160014b9861a8bf4af3ef5b96413d1c192912941e13ecc78c70100000000001600149cd3f14ce8775b771875d0826460e9255455aa8e78c70100000000001600143d264a7bfc3a3d3eaa205f9684be301da1729a3778c7010000000000160014cc74af1625a4f581dcd16bcb80ad9367958bf50d78c7010000000000160014b3738c4ad154dfbe32b43075eb01414b6206273a02473044022038884ba02e77e3c53c62412be065352b8c96b2420e117816e16b24fd6d9c47b202204257aad3ae0c9e42930ad2bb024b5adcda6f876c95b3772ae2ffa72a28327382012102449be5fb74725255eeeb50eba930fa87705f21e99d13cd710cf2c1f21153c8080247304402200a29d109643ddfb04fd26503c7e9887c9eabdc5f0ecc5072a357b91cd669ee8b022060c211a7c33eebcee308c641988e8cc67dbbe3b65856d4c69a2501a8694b0187012102131ef894942f0f9fe8f3455084c17f010549a7702bbfd74e4dff489738d0827f024730440220433cc9d973edfc3def7ec6afc49e18551e6a8af3267b75b6b72081bb80f99bbe022022473db462bcb6b698928c27fbda8bfe616cab07a62090e7dd69c4dd3a872523012103ac5d971ae3625f58c3ef5bcac46825653882bbb6a466f84d7226489dd876a1510247304402201f84bd62ab8e9986834da1f4e5e68c65d32cae7c6e2de69aa79666a53ea3696c022071a187d7248814080474e7c3a3fbbf771158f1e1c50e7d4e3c31a633f3af53fe012103025dc12c2024927ccbacb7f383d9a1a7d0709aaf73c7b22b794b82927f3be81a02473044022042636fb3f212af504b3196f27669c213342996f798aae6f50e4da0349f44c13e022076ce03586a0b6e387f330f6ef2ffc2fa952fca8c1391e0f68287ac9e17ff8381012103459928810bb656190ec36ca492f1e9c9ed84cc6ae09a0d2e9c62168bee2745b800000000"
}
```

**Comment by luke-jr** [comment]
@1440000bytes Did you provide it with the necessary foreign output metadata? (It's impossible to fix without that)

**Comment by 1440000bytes** [comment]
> @1440000bytes Did you provide it with the necessary foreign output metadata? (It's impossible to fix without that)
> 
> eg (check the outputs specified are what you intend)
> 
> ```
> bitcoin-cli -rpcwallet="w5" submitrawtransactiontowallet 02000000000105b4c6a8258717bce1b92f761acf76324bb808b78879ff3ccead1452de1ad5bf0a0000000000ffffffffa8f3781e6929b82fbe8a4e821497932266a2c051b20324baba0d6ae33ffddbf00100000000ffffffff3105dc25c8bd51e921aecda35b35ca0f8ab7239f5d3452c969ce75731fcdc0f00100000000ffffffffc0c6b85c082753edfd21e9fe286f6b62dbfc6cb30d61a92667263f4057f02d170000000000fffffffff9338d880fab8b9904dbdad9c51e737b50ca670c3db9b8df195fb03e18d48cdd0100000000ffffffff0578c7010000000000160014b9861a8bf4af3ef5b96413d1c192912941e13ecc78c70100000000001600149cd3f14ce8775b771875d0826460e9255455aa8e78c70100000000001600143d264a7bfc3a3d3eaa205f9684be301da1729a3778c7010000000000160014cc74af1625a4f581dcd16bcb80ad9367958bf50d78c7010000000000160014b3738c4ad154dfbe32b43075eb01414b6206273a02473044022038884ba02e77e3c53c62412be065352b8c96b2420e117816e16b24fd6d9c47b202204257aad3ae0c9e42930ad2bb024b5adcda6f876c95b3772ae2ffa72a28327382012102449be5fb74725255eeeb50eba930fa87705f21e99d13cd710cf2c1f21153c8080247304402200a29d109643ddfb04fd26503c7e9887c9eabdc5f0ecc5072a357b91cd669ee8b022060c211a7c33eebcee308c641988e8cc67dbbe3b65856d4c69a2501a8694b0187012102131ef894942f0f9fe8f3455084c17f010549a7702bbfd74e4dff489738d0827f024730440220433cc9d973edfc3def7ec6afc49e18551e6a8af3267b75b6b72081bb80f99bbe022022473db462bcb6b698928c27fbda8bfe616cab07a62090e7dd69c4dd3a872523012103ac5d971ae3625f58c3ef5bcac46825653882bbb6a466f84d7226489dd876a1510247304402201f84bd62ab8e9986834da1f4e5e68c65d32cae7c6e2de69aa79666a53ea3696c022071a187d7248814080474e7c3a3fbbf771158f1e1c50e7d4e3c31a633f3af53fe012103025dc12c2024927ccbacb7f383d9a1a7d0709aaf73c7b22b794b82927f3be81a02473044022042636fb3f212af504b3196f27669c213342996f798aae6f50e4da0349f44c13e022076ce03586a0b6e387f330f6ef2ffc2fa952fca8c1391e0f68287ac9e17ff8381012103459928810bb656190ec36ca492f1e9c9ed84cc6ae09a0d2e9c62168bee2745b800000000 '{"foreign_outputs": [0,1,2,3]}'
> ```

So I need to run `submitrawtransactiontowallet` for every coinjoin?

**Comment by luke-jr** [comment]
Yes

**Comment by 1440000bytes** [comment]
> Yes

This is still wrong in my case because w5 wallet did not send amount displayed in the json:

```
bitcoin-cli -rpcwallet="w5" submitrawtransactiontowallet 02000000000105b4c6a8258717bce1b92f761acf76324bb808b78879ff3ccead1452de1ad5bf0a0000000000ffffffffa8f3781e6929b82fbe8a4e821497932266a2c051b20324baba0d6ae33ffddbf00100000000ffffffff3105dc25c8bd51e921aecda35b35ca0f8ab7239f5d3452c969ce75731fcdc0f00100000000ffffffffc0c6b85c082753edfd21e9fe286f6b62dbfc6cb30d61a92667263f4057f02d170000000000fffffffff9338d880fab8b9904dbdad9c51e737b50ca670c3db9b8df195fb03e18d48cdd0100000000ffffffff0578c7010000000000160014b9861a8bf4af3ef5b96413d1c192912941e13ecc78c70100000000001600149cd3f14ce8775b771875d0826460e9255455aa8e78c70100000000001600143d264a7bfc3a3d3eaa205f9684be301da1729a3778c7010000000000160014cc74af1625a4f581dcd16bcb80ad9367958bf50d78c7010000000000160014b3738c4ad154dfbe32b43075eb01414b6206273a02473044022038884ba02e77e3c53c62412be065352b8c96b2420e117816e16b24fd6d9c47b202204257aad3ae0c9e42930ad2bb024b5adcda6f876c95b3772ae2ffa72a28327382012102449be5fb74725255eeeb50eba930fa87705f21e99d13cd710cf2c1f21153c8080247304402200a29d109643ddfb04fd26503c7e9887c9eabdc5f0ecc5072a357b91cd669ee8b022060c211a7c33eebcee308c641988e8cc67dbbe3b65856d4c69a2501a8694b0187012102131ef894942f0f9fe8f3455084c17f010549a7702bbfd74e4dff489738d0827f024730440220433cc9d973edfc3def7ec6afc49e18551e6a8af3267b75b6b72081bb80f99bbe022022473db462bcb6b698928c27fbda8bfe616cab07a62090e7dd69c4dd3a872523012103ac5d971ae3625f58c3ef5bcac46825653882bbb6a466f84d7226489dd876a1510247304402201f84bd62ab8e9986834da1f4e5e68c65d32cae7c6e2de69aa79666a53ea3696c022071a187d7248814080474e7c3a3fbbf771158f1e1c50e7d4e3c31a633f3af53fe012103025dc12c2024927ccbacb7f383d9a1a7d0709aaf73c7b22b794b82927f3be81a02473044022042636fb3f212af504b3196f27669c213342996f798aae6f50e4da0349f44c13e022076ce03586a0b6e387f330f6ef2ffc2fa952fca8c1391e0f68287ac9e17ff8381012103459928810bb656190ec36ca492f1e9c9ed84cc6ae09a0d2e9c62168bee2745b800000000 '{"foreign_outputs": [0,1,2,3]}'

$ bitcoin-cli -rpcwallet="w5" gettransaction 75e490b10b15a6a0422f25ff66ad98ef70390c8fecaac02712705dce8cc3564b
{
  "amount": -0.00583000,
  "fee": 0.00465000,
  "confirmations": 2177,
  "blockhash": "00000007567d87fadb4673d0cb876cb1b7b9dbfedc1d9a0e40af7380d9da8435",
  "blockheight": 104271,
  "blockindex": 1,
  "blocktime": 1660978126,
  "txid": "75e490b10b15a6a0422f25ff66ad98ef70390c8fecaac02712705dce8cc3564b",
  "wtxid": "e4b91ce21961a9e5498e2ff396f5e09fa8c7f3a36ef77abd4207f71855599990",
  "walletconflicts": [
  ],
  "time": 1660976747,
  "timereceived": 1660976747,
  "bip125-replaceable": "no",
  "details": [
    {
      "address": "tb1qkdeccjk32n0muv45xp67kq2pfd3qvfe6wym0ey",
      "category": "send",
      "amount": -0.00116600,
      "vout": 4,
      "fee": -0.00001400,
      "abandoned": false
    }
  ],
  "hex": "02000000000105b4c6a8258717bce1b92f761acf76324bb808b78879ff3ccead1452de1ad5bf0a0000000000ffffffffa8f3781e6929b82fbe8a4e821497932266a2c051b20324baba0d6ae33ffddbf00100000000ffffffff3105dc25c8bd51e921aecda35b35ca0f8ab7239f5d3452c969ce75731fcdc0f00100000000ffffffffc0c6b85c082753edfd21e9fe286f6b62dbfc6cb30d61a92667263f4057f02d170000000000fffffffff9338d880fab8b9904dbdad9c51e737b50ca670c3db9b8df195fb03e18d48cdd0100000000ffffffff0578c7010000000000160014b9861a8bf4af3ef5b96413d1c192912941e13ecc78c70100000000001600149cd3f14ce8775b771875d0826460e9255455aa8e78c70100000000001600143d264a7bfc3a3d3eaa205f9684be301da1729a3778c7010000000000160014cc74af1625a4f581dcd16bcb80ad9367958bf50d78c7010000000000160014b3738c4ad154dfbe32b43075eb01414b6206273a02473044022038884ba02e77e3c53c62412be065352b8c96b2420e117816e16b24fd6d9c47b202204257aad3ae0c9e42930ad2bb024b5adcda6f876c95b3772ae2ffa72a28327382012102449be5fb74725255eeeb50eba930fa87705f21e99d13cd710cf2c1f21153c8080247304402200a29d109643ddfb04fd26503c7e9887c9eabdc5f0ecc5072a357b91cd669ee8b022060c211a7c33eebcee308c641988e8cc67dbbe3b65856d4c69a2501a8694b0187012102131ef894942f0f9fe8f3455084c17f010549a7702bbfd74e4dff489738d0827f024730440220433cc9d973edfc3def7ec6afc49e18551e6a8af3267b75b6b72081bb80f99bbe022022473db462bcb6b698928c27fbda8bfe616cab07a62090e7dd69c4dd3a872523012103ac5d971ae3625f58c3ef5bcac46825653882bbb6a466f84d7226489dd876a1510247304402201f84bd62ab8e9986834da1f4e5e68c65d32cae7c6e2de69aa79666a53ea3696c022071a187d7248814080474e7c3a3fbbf771158f1e1c50e7d4e3c31a633f3af53fe012103025dc12c2024927ccbacb7f383d9a1a7d0709aaf73c7b22b794b82927f3be81a02473044022042636fb3f212af504b3196f27669c213342996f798aae6f50e4da0349f44c13e022076ce03586a0b6e387f330f6ef2ffc2fa952fca8c1391e0f68287ac9e17ff8381012103459928810bb656190ec36ca492f1e9c9ed84cc6ae09a0d2e9c62168bee2745b800000000"
}
```

**Comment by luke-jr** [comment]
Then you provided the wrong metadata, because that looks right to me?

**Comment by 1440000bytes** [comment]
> Then you provided the wrong metadata, because that looks right to me?

I think it was correct output but i need a break for a few days before reviewing anything

**Comment by Riahiamirreza** [comment]
Would you give me a hint about what does `CachedTxGetAmounts` function do? There is no comments around to explain more about it.

File: src/wallet/rpc/spend.cpp [hunk 1]
```diff
@@ -1457,6 +1457,102 @@ RPCHelpMan sendall()
...
+        [&](const RPCHelpMan& self, const JSONRPCRequest& request) -> UniValue
+        {
+            std::shared_ptr<CWallet> const pwallet{GetWalletForJSONRPCRequest(request)};
+            if (!pwallet) return UniValue::VNULL;
+
+            RPCTypeCheck(request.params, {
+                UniValue::VSTR, // hexstring
+                UniValue::VOBJ, // options
+            }, /*fAllowNull=*/true);
+
```
Comment (amovfx) [inline src/wallet/rpc/spend.cpp:1497]: Would this PR be of interest to you?
https://github.com/bitcoin/bitcoin/pull/26039

**Review by amovfx — COMMENTED** [review]
This was an excellent pr for the review club. I learned a ton from this.

Is this worth testing with the legacy wallet?

File: src/wallet/transaction.h [hunk 2]
```diff
@@ -231,6 +241,15 @@ class CWalletTx
...
         }
+        if (!m_foreign_outputs.empty()) {
+            std::string s(m_foreign_outputs.size(), 0);
+            for (auto i = m_foreign_outputs.size(); i--; ) {
+                if (m_foreign_outputs.at(i)) {
+                    s[i / 8] |= 1 << (i % 8);
+                }
+            }
+            mapValueCopy["fout"] = s;
+        }
```
Comment (amovfx) [inline src/wallet/transaction.h:252]: I am really looking forward to learning what this does. For the life of me I can't figure out what you are doing with the string or why this works.

File: src/wallet/rpc/spend.cpp [hunk 3]
```diff
@@ -1457,6 +1457,102 @@ RPCHelpMan sendall()
...
+                RPCTypeCheckObj(options, {
+                    {"foreign_outputs", UniValueType(UniValue::VARR)},
+                }, /*fAllowNull=*/true, /*fStrict=*/true);
+
+                if (!options["foreign_outputs"].isNull()) {
+                    auto output_count = tx->vout.size();
+                    foreign_outputs.emplace(output_count);
+                    const auto& foreign_outputs_uv = options["foreign_outputs"].get_array();
+                    size_t final_foreign_outputs_size = 0;
+                    for (auto i = foreign_outputs_uv.size(); i--; ) {
```
Comment (MarcoFalke) [inline src/wallet/rpc/spend.cpp:1516]: node0 stderr wallet/rpc/spend.cpp:1516:63: runtime error: unsigned integer overflow: 0 - 1 cannot be represented in type 'size_t' (aka 'unsigned long')

Comment (luke-jr) [reply src/wallet/rpc/spend.cpp:1516]: AFAIK this is defined behaviour and correct.

File: src/wallet/transaction.h (hunk 2 above)
Comment (amovfx) [reply src/wallet/transaction.h:252]: [I got an answer on bitcoin stack overflow.](https://bitcoin.stackexchange.com/questions/115208/help-explaining-some-c-code-i-havent-seen-before)

File: src/wallet/transaction.h [hunk 4]
```diff
@@ -231,6 +241,15 @@ class CWalletTx
         if (nTimeSmart) {
             mapValueCopy["timesmart"] = strprintf("%u", nTimeSmart);
         }
+        if (!m_foreign_outputs.empty()) {
+            std::string s(m_foreign_outputs.size(), 0);
```
Comment (LarryRuane) [inline src/wallet/transaction.h:245]:
```suggestion
            std::string s((m_foreign_outputs.size() + 7) / 8, 0);
```

**Review by crimscim — APPROVED** [review]

**Review by stickies-v — COMMENTED** [review]
Did a first code review already. If we want to track foreign outputs, this approach makes sense to me.

I'm wondering, would it be more user-friendly to instead track non-foreign outputs? For a user that is only using Core to create transactions, we could automatically track all outputs created internally and assume everything that's not in that set is foreign. This way, the user would not have to submit any additional input for `listtransactions` to properly work? And could still allow users to manually set outputs to be non-foreign, similar to what's implemented now? I'm not very familiar with the wallet, so apologies if I'm missing something obvious.

I think it would also be nice to update the PR description with a bit more context on what the importance of this PR is, i.e. what is the impact of the current non-sensical behaviour. Are there functional implications (I'm not sure what the general purpose of `listtransactions` is)?

File: test/functional/wallet_listtransactions.py [hunk 5]
```diff
@@ -277,6 +278,79 @@ def normalize_list(txs):
...
+        self.nodes[1].sendrawtransaction(rawtx, 0)
+
+        # Currently, the result is nonsensical/broken:
+        # node 0 thinks it received 1 BTC, then sent 2 BTC + 47 BTC change + 1 BTC with a -47 BTC fee
+        # node 1 thinks it received 2 BTC, then sent 2 BTC + 47 BTC change + 1 BTC with a -47 BTC fee
+        txlist_node0 = self.nodes[0].listtransactions('*', 5)
+        txlist_node1 = self.nodes[1].listtransactions('*', 5)
+        assert txlist_node0.pop(0)['txid'] != txid
+        assert txlist_node1.pop(0)['txid'] != txid
+        txlist = list((tx['address'], tx['amount'], tx['category'], tx.get('fee', None)) for tx in txlist_node0 + txlist_node1)
```
[the remark this answers is not in the document]
Comment (luke-jr) [reply test/functional/wallet_listtransactions.py:320]: I think explicitly using `list(...)` is clearer (`[...]` seems to work, but it's not obvious that it isn't just making a list with a single generator element)

File: test/functional/wallet_listtransactions.py [hunk 6]
```diff
@@ -277,6 +278,88 @@ def normalize_list(txs):
...
+            assert txlist_node0.pop(0)['txid'] != txid
+            assert txlist_node1.pop(0)['txid'] != txid
+            txlist = list((tx['address'], tx['amount'], tx['category'], tx.get('fee', None)) for tx in txlist_node0 + txlist_node1)
+            assert_equal(txlist, [
+                # node 0:
+                (addr_pay0, 1, 'receive', None),
+                (addr_pay1, -2, 'send', -1),
+                # node 1:
+                (addr_pay1, 2, 'receive', None),
+                (addr_pay0, -1, 'send', -2),
```
[the remark this answers is not in the document]
Comment (luke-jr) [reply test/functional/wallet_listtransactions.py:352]: They're paying each other, not themselves.

File: src/wallet/receive.cpp [hunk 7]
```diff
@@ -202,11 +202,7 @@ void CachedTxGetAmounts(const CWallet& wallet, const CWalletTx& wtx,
 
     // Compute fee:
     CAmount nDebit = CachedTxGetDebit(wallet, wtx, filter);
-    if (nDebit > 0) // debit>0 means we signed/sent this transaction
-    {
-        CAmount nValueOut = wtx.tx->GetValueOut();
```
[the remark this answers is not in the document]
Comment (luke-jr) [reply src/wallet/receive.cpp:207]: I don't see where `CachedTxGetDebit` calls `MoneyRange` on outputs, but in any case, it's not a useful check here.

File: src/wallet/rpc/spend.cpp [hunk 8]
```diff
@@ -1457,6 +1457,105 @@ RPCHelpMan sendall()
...
+
+            std::optional<std::vector<bool>> foreign_outputs;
+            if (!request.params[1].isNull()) {
+                const UniValue& options = request.params[1];
+                RPCTypeCheckObj(options, {
+                    {"foreign_outputs", UniValueType(UniValue::VARR)},
+                }, /*fAllowNull=*/true, /*fStrict=*/true);
+
+                if (!options["foreign_outputs"].isNull()) {
+                    auto output_count = tx->vout.size();
```
[the remark this answers is not in the document]
Comment (luke-jr) [reply src/wallet/rpc/spend.cpp:1512]: I don't really care, so fine, but IMO it's less readable. Is there a reason for it, that should perhaps get documented?

File: src/wallet/rpc/spend.cpp [hunk 9]
```diff
@@ -1457,6 +1457,105 @@ RPCHelpMan sendall()
...
+                        },
+                    },
+                },
+            },
+        },
+        RPCResult{RPCResult::Type::NONE, "", ""},
+        RPCExamples{
+            "\nCreate a transaction\n"
+            + HelpExampleCli("createrawtransaction", "\"[{\\\"txid\\\" : \\\"mytxid\\\",\\\"vout\\\":0}]\" \"{\\\"myaddress\\\":0.01}\"") +
+            "Sign the transaction, and get back the hex\n"
```
[the remark this answers is not in the document]
Comment (luke-jr) [reply src/wallet/rpc/spend.cpp:1481]: Done

File: test/functional/wallet_listtransactions.py [hunk 10]
```diff
@@ -277,6 +278,88 @@ def normalize_list(txs):
...
+            ])
+
+        final_check()
+
+        # Ensure metadata is preserved across restart
+        self.nodes[0].args = ["valgrind"] +self.nodes[0].args
```
[the remark this answers is not in the document]
Comment (luke-jr) [reply test/functional/wallet_listtransactions.py:358]: Accident leftover from debugging (I couldn't get `--valgrind` to work at all). Removed.

**Comment by 1440000bytes** [comment]
I took a brook. Although I think this PR needs to consider what is a "conjoin transaction"

File: src/wallet/rpc/spend.cpp (hunk 3 above)
Comment (luke-jr) [reply src/wallet/rpc/spend.cpp]: This one didn't need the index at all, so changed it.

Opened https://github.com/llvm/llvm-project/issues/57794 to address the false positives directly, and added exceptions for the others for now.

**Review by Tetita — APPROVED** [review]

File: src/wallet/wallet.h [hunk 11]
```diff
@@ -513,6 +513,7 @@ class CWallet final : public WalletStorage, public interfaces::Chain::Notificati
      * @return the recently added wtx pointer or nullptr if there was a db write error.
      */
     CWalletTx* AddToWallet(CTransactionRef tx, const TxState& state, const UpdateWalletTxFn& update_wtx=nullptr, bool fFlushOnClose=true, bool rescanning_old_block = false);
+    bool WriteTx(CWalletTx&);
```
[the remark this answers is not in the document]
Comment (Tetita) [reply src/wallet/wallet.h:516]: #25991

File: test/sanitizer_suppressions/ubsan [hunk 12]
```diff
@@ -53,6 +53,9 @@ unsigned-integer-overflow:policy/fees.cpp
 unsigned-integer-overflow:prevector.h
 unsigned-integer-overflow:script/interpreter.cpp
 unsigned-integer-overflow:txmempool.cpp
+unsigned-integer-overflow:wallet/transaction.cpp
+unsigned-integer-overflow:wallet/transaction.h
+unsigned-integer-overflow:wallet/wallet.cpp
```
[the remark this answers is not in the document]
Comment (Tetita) [reply test/sanitizer_suppressions/ubsan:58]: https://github.com/bitcoin/bitcoin/pull/25991#discussion_r973550983

**Comment by 1440000bytes** [comment]
> I took a break. Although I think this PR needs to consider what is a "conjoin transaction"

Sorry for sharing less details. There are 3 main coinjoin implementations available but they have different type of transactions and its to spot on-chain. Does this PR work for all coinjoin that do not follow any of these implementations?

File: src/wallet/receive.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (stickies-v) [reply src/wallet/receive.cpp:207]: `CachedTxGetDebit` calls `GetCachableAmount` which calls `GetDebit` which calls `MoneyRange`.

**Comment by luke-jr** [comment]
> There are 3 main coinjoin implementations available but they have different type of transactions and its easy to spot on-chain. Does this PR work for all coinjoin that do not follow any of these implementations?

This PR is neutral about how you create the transaction. It just provides a way to tell this wallet about the details needed to display it correctly (and acts on that, for RPC).

**Comment by luke-jr** [comment]
Added a wallet flag to prevent old wallets from trying to load ones with foreign-output metadata. (Currently a one-way flag: it isn't possible to unset, even if you remove foreign-output metadata later - it would need to scan your entire wallet to know to remove it)

File: src/wallet/receive.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (luke-jr) [reply src/wallet/receive.cpp:207]: `GetDebit` works with inputs, not outputs.

**Comment by achow101** [comment]
Closing as up for grabs due to lack of activity.

**Pull request closed without merging**